pandas/core/generic.py

```diff
@@ -4880,20 +4880,39 @@ def describe(self, percentiles=None, include=None, exclude=None):
 
         # median should always be included
         if (percentiles != 0.5).all():  # median isn't included
-            lh = percentiles[percentiles < .5]
-            uh = percentiles[percentiles > .5]
-            percentiles = np.hstack([lh, 0.5, uh])
-
-        def pretty_name(x):
-            x *= 100
-            if x == int(x):
-                return '%.0f%%' % x
-            else:
-                return '%.1f%%' % x
+            percentiles = np.concatenate((percentiles, [0.5]))
+
+        # sort and check for duplicates
+        p_len = len(percentiles)
+        percentiles = np.unique(percentiles)
+        if len(percentiles) < p_len:
+            raise ValueError("percentiles cannot contain duplicates")
+
+        def prettify(percentiles):
+            pcts = 100 * percentiles
+            int_idx = (pcts.astype(int) == pcts)
+
+            if np.all(int_idx):
+                out = pcts.astype(int).astype(str)
+                return [i + '%' for i in out]
+
+            to_begin = pcts[0] if pcts[0] > 0 else None
+            to_end = 100 - pcts[-1] if pcts[-1] < 100 else None
+            # Least precision that keeps percentiles unique after rounding
+            prec = -np.floor(np.log10(np.min(
+                np.ediff1d(pcts, to_begin=to_begin, to_end=to_end)
+            ))).astype(int)
+            prec = max(1, prec)
+            out = np.empty_like(pcts, dtype=object)
+            out[int_idx] = pcts[int_idx].astype(int).astype(str)
+            out[~int_idx] = pcts[~int_idx].round(prec).astype(str)
+            return [i + '%' for i in out]
+
+        pretty_percentiles = prettify(percentiles)
```

Comment:
I wasn't very clear about the "ugliness" - see my comment to `describe_numeric_1d`.

Anyway, I don't mind the function `prettify` is here. But if you think it could be useful for other purposes, I can make it more general without efficiency loss in this case.